test/integration/data/elastic_torch_main.py

```diff
@@ -0,0 +1,136 @@
+# Copyright 2020 Uber Technologies, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+from __future__ import absolute_import
+
+import argparse
+import json
+import os
+import time
+
+import torch
+import torch.nn.functional as F
+
+import horovod.torch as hvd
+
+parser = argparse.ArgumentParser(description='PyTorch Elastic Test',
+                                 formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+
+parser.add_argument('--batches-per-epoch', type=int, default=10,
+                    help='number of batches per epoch')
+parser.add_argument('--batches-per-commit', type=int, default=1,
+                    help='number of batches per commit of the elastic state object')
+parser.add_argument('--epochs', type=int, default=3,
+                    help='number of epochs')
+parser.add_argument('--logfile', default='/tmp/logfile.txt',
+                    help='log file to record results (one line per epoch)')
+parser.add_argument('--discovery-schedule', default='[]',
+                    help='JSON string specifying schedule of host updates each epoch')
+parser.add_argument('--exit-schedule',
+                    help='JSON string mapping from (epoch, batch) to list of ranks to exit at that time')
+
+args = parser.parse_args()
+
+hvd.init()
+
+batch_size = 32
+data = torch.randn(batch_size, 2)
+target = torch.LongTensor(batch_size).random_() % 2
+
+lr = 0.001
+model = torch.nn.Sequential(torch.nn.Linear(2, 2))
+optimizer = torch.optim.SGD(model.parameters(), lr=lr * hvd.size())
+optimizer = hvd.DistributedOptimizer(optimizer, named_parameters=model.named_parameters())
+
+hostname = os.environ.get('HOROVOD_HOSTNAME')
+start_rank = int(os.environ.get('HOROVOD_RANK', 0))
+
+discovery_schedule = json.loads(args.discovery_schedule)
+epoch_to_hosts = {epoch: hosts for epoch, hosts in discovery_schedule if epoch is not None}
+default_hosts = discovery_schedule[-1][1] if len(discovery_schedule) > 0 else []
+
+exit_schedule = json.loads(args.exit_schedule) if args.exit_schedule else {}
+
+
+def check_exit(epoch, batch):
+    key = str((epoch, batch))
+    if key in exit_schedule:
+        ranks_to_exit = exit_schedule[key]
+        if start_rank in ranks_to_exit:
+            raise RuntimeError('check_rank and exit epoch={} batch={} start_rank={} rank={}'
```

Comment:
This only tests an exception in the training loop but not a lost worker node. Both cases should be tested, a new `--exit-mode=[exception|kill]` could be added.

Changing this to `psutil.Process(os.getpid()).kill()` makes the elastic horovod job fail and not recover. The other workers terminate on a `gloo::IoException`:
```
Mon May 11 11:44:15 2020[0]<stderr>:Killed
Mon May 11 11:44:15 2020[3]<stderr>:terminate called after throwing an instance of 'gloo::IoException'
Mon May 11 11:44:15 2020[3]<stderr>:  what():  [horovod/third_party/gloo/gloo/transport/tcp/pair.cc:563] Read error [127.0.0.1]:16802: Connection reset by peer
```

Calling `sys.exit(1)` does not make the horovod job fail because some graceful shutdown is still happening.

Reply:
Good catch.  The problem here was that we were not properly handling failures in the control plane.  Should be fixed now.